types/jest/index.d.ts

```diff
@@ -693,67 +693,56 @@ declare namespace jest {
         getState(): MatcherState & Record<string, any>;
     }
 
-    type JestMatchers<T> = JestMatchersShape<Matchers<void, T>, Matchers<Promise<void>, T>>;
+    type PromiseOf<T> = T extends Promise<infer S> ? S : never;
+    type JestMatchers<T> = JestMatchersShape<Matchers<void, T>, Matchers<Promise<void>, PromiseOf<T>>, Matchers<Promise<void>, any>>;
 
-    type JestMatchersShape<TNonPromise extends {} = {}, TPromise extends {} = {}> = {
+    type JestMatchersShape<TNonPromise extends {} = {}, TResolve extends {} = {}, TReject extends {} = {}> = {
         /**
          * Use resolves to unwrap the value of a fulfilled promise so any other
          * matcher can be chained. If the promise is rejected the assertion fails.
          */
-        resolves: AndNot<TPromise>,
+        resolves: AndNot<TResolve>,
         /**
          * Unwraps the reason of a rejected promise so any other matcher can be chained.
          * If the promise is fulfilled the assertion fails.
          */
-        rejects: AndNot<TPromise>
+        rejects: AndNot<TReject>
     } & AndNot<TNonPromise>;
     type AndNot<T> = T & {
         not: T
     };
 
+    type JasmineMatcher =
+        | ReturnType<typeof jasmine.any>
+        | ReturnType<typeof jasmine.arrayContaining>
+        | ReturnType<typeof jasmine.objectContaining>;
+
     // should be R extends void|Promise<void> but getting dtslint error
     interface Matchers<R, T = {}> {
         /**
          * Ensures the last call to a mock function was provided specific args.
-         *
-         * Optionally, you can provide a type for the expected arguments via a generic.
-         * Note that the type must be either an array or a tuple.
          */
-        // tslint:disable-next-line: no-unnecessary-generics
-        lastCalledWith<E extends any[]>(...args: E): R;
+        lastCalledWith(...args: T extends (...args: any) => any ? Parameters<T> : any[]): R;
```

Comment:
These are all breaking changes for anyone who explicitly set E

Reply:
Yes, I thought it was worth the cost of the huge benefit for those who don't use it and the huge manual labor that was involved. 

I've found another safe option by using extends now. I think this should be fixed now.